packages/grafana-ui/src/components/uPlot/utils.ts

```diff
@@ -41,115 +40,219 @@ interface StackMeta {
-): AlignedData {
-  const frame = frames[0];
-  const result: any[] = [];
-  const stackingGroups: Map<string, number[]> = new Map();
-  let seriesIndex = 0;
-
-  for (let i = 0; i < frame.fields.length; i++) {
-    const f = frame.fields[i];
-
-    if (f.type === FieldType.time) {
-      result.push(ensureTimeField(f).values.toArray());
-      seriesIndex++;
-      continue;
+export interface StackingGroup {
+  series: number[];
+  dir: number;
+}
+
+// generates bands between adjacent group series
+/** @internal */
+export function getStackingBands(group: StackingGroup) {
+  let bands: uPlot.Band[] = [];
+  let { series, dir } = group;
+  let lastIdx = series.length - 1;
+
+  let rSeries = series.slice().reverse();
+
+  rSeries.forEach((si, i) => {
+    if (i !== lastIdx) {
+      let nextIdx = rSeries[i + 1];
+      bands.push({
+        series: [si, nextIdx],
+        // fill direction is inverted from stack direction
+        dir: (-1 * dir) as 1 | -1,
+      });
     }
+  });
 
-    collectStackingGroups(f, stackingGroups, seriesIndex);
-    const customConfig: GraphFieldConfig = f.config.custom || {};
+  return bands;
+}
 
-    const values = f.values.toArray();
+// expects an AlignedFrame
+/** @internal */
+export function getStackingGroups(frame: DataFrame) {
+  let groups: Map<string, StackingGroup> = new Map();
 
-    if (customConfig.transform === GraphTransform.NegativeY) {
-      result.push(values.map((v) => (v == null ? v : v * -1)));
-    } else if (customConfig.transform === GraphTransform.Constant) {
-      result.push(new Array(values.length).fill(values[0]));
-    } else {
-      result.push(values);
+  frame.fields.forEach(({ config, values }, i) => {
+    // skip x or time field
+    if (i === 0) {
+      return;
+    }
+
+    let { custom } = config;
+
+    if (custom == null) {
+      return;
+    }
+
+    // TODO: currently all AlignedFrame fields end up in uplot series & data, even custom.hideFrom?.viz
+    // ideally hideFrom.viz fields would be excluded so we can remove this
+    if (custom.hideFrom?.viz) {
+      return;
+    }
+
+    let { stacking } = custom;
+
+    if (stacking == null) {
+      return;
+    }
+
+    let { mode: stackingMode, group: stackingGroup } = stacking;
+
+    // not stacking
+    if (stackingMode === StackingMode.None) {
+      return;
+    }
+
+    // will this be stacked up or down after any transforms applied
+    let vals = values.toArray();
+    let transform = custom.transform;
+    let stackDir =
+      transform === GraphTransform.Constant
+        ? vals[0] > 0
+          ? 1
+          : -1
+        : transform === GraphTransform.NegativeY
+        ? vals.some((v) => v > 0)
+          ? -1
+          : 1
+        : vals.some((v) => v > 0)
+        ? 1
+        : -1;
+
+    let drawStyle = custom.drawStyle;
+    let drawStyle2 =
+      drawStyle === GraphDrawStyle.Bars
+        ? custom.barAlignment
+        : drawStyle === GraphDrawStyle.Line
+        ? custom.lineInterpolation
```

Comment:
this will prevent stacking of line series with different interpolations (e.g. linear + smooth + stepped). and will auto-group stacks based on interpolation, in addition to other things like scaleKey, etc.

generally i think this is the right thing to do in 99% of the cases. we can always relax this later if people present compelling use cases.